### PR TITLE
Export the Skia ShadowMaskFilter API

### PR DESCRIPTION
**Pull request by jason-simmons**
See https://github.com/flutter/flutter/issues/6807

### CODE REVIEW

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -861,9 +861,27 @@ class MaskFilter extends NativeFieldWrapperClass2 {
...
-  void _constructor(int style, double sigma) native "MaskFilter_constructor";
+  void _constructorBlur(int style, double sigma) native "MaskFilter_constructorBlur";
+
+  MaskFilter.shadow(double occluderHeight,
+                    double lightPosX,
+                    double lightPosY,
+                    double lightPosZ,
+                    double lightRadius,
+                    double ambientAlpha,
+                    double spotAlpha) {
```
Comment (abarth) [inline lib/ui/painting.dart]: We should make these named arguments.  That will make the call sites much more readable.

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -861,9 +861,27 @@ class MaskFilter extends NativeFieldWrapperClass2 {
...
+    _constructorShadow(occluderHeight, lightPosX, lightPosY, lightPosZ,
+                       lightRadius, ambientAlpha, spotAlpha);
+  }
+  void _constructorShadow(double occluderHeight,
+                          double lightPosX,
+                          double lightPosY,
+                          double lightPosZ,
+                          double lightRadius,
+                          double ambientAlpha,
+                          double spotAlpha) native "MaskFilter_constructorShadow";
```
Comment (abarth) [inline lib/ui/painting.dart]: These can stay as positional.

File: lib/ui/painting/mask_filter.cc [hunk 3]
```diff
@@ -11,26 +11,54 @@
...
+                                                 double lightPosZ,
+                                                 double lightRadius,
+                                                 double ambientAlpha,
+                                                 double spotAlpha) {
+  if (!(occluderHeight > 0 &&
+        lightPosZ > occluderHeight &&
+        lightRadius > 0 &&
+        ambientAlpha >= 0 &&
+        spotAlpha >= 0))
+    Dart_ThrowException(tonic::ToDart("Invalid shadow mask filter parameters"));
```
Comment (abarth) [inline lib/ui/painting/mask_filter.cc]: We usually check these invariants in Dart.

File: lib/ui/painting.dart [hunk 4]
```diff
@@ -861,9 +861,27 @@ class MaskFilter extends NativeFieldWrapperClass2 {
   ///
   /// A blur is an expensive operation and should therefore be used sparingly.
   MaskFilter.blur(BlurStyle style, double sigma) {
-    _constructor(style.index, sigma);
+    _constructorBlur(style.index, sigma);
   }
-  void _constructor(int style, double sigma) native "MaskFilter_constructor";
+  void _constructorBlur(int style, double sigma) native "MaskFilter_constructorBlur";
+
+  MaskFilter.shadow(double occluderHeight,
```
Comment (abarth) [inline lib/ui/painting.dart]: Needs docs.

**Comment by abarth** [comment]
LGTM

... with docs

**Comment by jason-simmons** [comment]
Updated the interface and added docs based on the Skia headers

File: lib/ui/painting.dart [hunk 5]
```diff
@@ -861,9 +861,47 @@ class MaskFilter extends NativeFieldWrapperClass2 {
...
+  /// `ambientAlpha`: Base opacity of the ambient occlusion shadow.
+  ///
+  /// `spotAlpha`: Base opacity of the displaced spot shadow.
+  MaskFilter.shadow({double occluderHeight: 0.0,
+                     double lightPosX: 0.0,
+                     double lightPosY: 0.0,
+                     double lightPosZ: 0.0,
+                     double lightRadius: 0.0,
+                     double ambientAlpha: 0.0,
+                     double spotAlpha: 0.0}) {
```
Comment (abarth) [inline lib/ui/painting.dart]: nit: Our house style would be to have a newline after the `{` and then to indent the next line by two spaces more than the MaskFilter.shadow line.  We'd also a newline before the last `}` and indent the line with the `}` to match the MaskFilter.shadow line.

**Comment by abarth** [comment]
LGTM

**Pull request merged**